nx/lib/nx/tensor.ex

```diff
@@ -143,8 +143,47 @@ defmodule Nx.Tensor do
   end
 
   @impl true
-  def get_and_update(_tensor, _index, _update) do
-    raise "Access.get_and_update/3 is not yet supported by Nx.Tensor"
+  def get_and_update(%Nx.Tensor{shape: {}} = tensor, _index, _update) do
+    raise ArgumentError,
+          "cannot use the tensor[index] syntax on scalar tensor #{inspect(tensor)}"
+  end
+
+  def get_and_update(tensor, index, update) do
+    if index < 0 or index >= elem(tensor.shape, 0) do
+      raise ArgumentError,
+            "index #{index} is out of bounds for axis 0 in shape #{inspect(tensor.shape)}"
+    end
+
+    {:ok, current_value} = fetch(tensor, index)
+    {_, new_value} = update.(current_value)
+
+    if current_value.shape != new_value.shape do
+      raise ArgumentError,
+            "the shape #{inspect(new_value.shape)} of the new value does not match the current value shape #{inspect(current_value.shape)}"
+    end
+
+    wrapper_shape =
+      tensor.shape
+      |> Tuple.delete_at(0)
+      |> Tuple.insert_at(0, 1)
+
+    new_value_wrapped = Nx.broadcast(new_value, wrapper_shape)
+
+    indices = get_slice_indices(index, tensor.shape)
+
+    impl = Nx.Shared.impl!(tensor)
+
+    updated_tensor =
+      tensor
+      |> impl.put_slice(tensor, indices, new_value_wrapped)
```

Comment:
```suggestion
    updated_tensor = Nx.put_slice(tensor, indices, new_value_wrapped)
```

I don't think you need to go below the public API for this

Reply:
For this I was taking as an example how the slices are generated in `fetch`, which uses the implementation slice directly. What are the main tradeoffs between one and other?

Reply:
put_slice will attempt to normalize the data but we have already normalize the data here. That's why we skipped. I think skipping here is fine!